### PR TITLE
Encode httpStatusCodes in error objects

### PR DESCRIPTION
**Pull request by ryanb8**
This PR allows for http status codes and raw error responses to be passed through to error objects.

An example usage:
```R
library(paws)
s3_svc <- s3()
b1 <- tryCatch(s3_svc$list_objects_v2(Bucket = "alskfjalskfjalkfj", MaxKeys =2), error = function(e) e)
b1$message
# > "NoSuchBucket (HTTP 404). The specified bucket does not exist"

attributes(b1)

# Output:
#> $names
#> [1] "message"
#> $class
#> [1] "http_404"   "http_400"   "http_error" "error"      "condition" 
#> $statusCode
#> [1] 404
#> 
#> $errorResponse
#> $errorResponse$Code
#> [1] "NoSuchBucket"
#> $errorResponse$Message
#> [1] "The specified bucket does not exist"
#> $errorResponse$BucketName
#> [1] "alskfjalskfjalkfj"
#> $errorResponse$RequestId
#> [1] "356852C376199D04"
#> $errorResponse$HostId
#> [1] "P2te81nv0wg9q1hoXsdKxUEFzwhB+JCLYS599APIKGLKpn8X/LaOioBS/a229v93uYOMXkFehIc="
```

This change is driven by the removal of the `format_error` and the introduction of it's replacement [`http_error`](https://github.com/ryanb8/paws/blob/e8705c61a6928027fb10b3265c2ae87d058ea23c/paws.common/R/error.R#L15-L63). Instead of just generating a string that is passed to stop, we generate a condition that inherits from error and respective http classes. We attach the raw response and the numeric httpStatusCode as attributes on the error object. This is heavily inspired by [httr's error handling](https://github.com/r-lib/httr/blob/master/R/response-status.r#L310)

Tests have been added for all handlers.

Some things that I definitely want y'all to weigh in on:
1. Naming Conventions - snake_case vs camelCase vs other
2. Should I add an `examples/error.R` file to demonstrate how to use this feature?

--------
Manual Tests I Used to Validate:

```R
#ryan_manual_tests

library(paws)

# s3_custom
s3_svc <- s3()
b1 <- tryCatch(s3_svc$list_objects_v2(Bucket = "alskfjalskfjalkfj", MaxKeys =2), error = function(e) e)
b1$message
class(b1)
attributes(b1)

# ec2 custom
# https://raw.githubusercontent.com/aws/aws-sdk-js/master/apis/ec2-2016-11-15.normal.json
ec2_svc <- ec2()
b2 <- tryCatch(ec2_svc$describe_hosts(HostIds = list("aslkfjalskfjalksfjalskfj")), error = function(e) e)
b2$message
class(b2)
attributes(b2)

# rest-json - Lambda
# https://github.com/aws/aws-sdk-js/blob/master/apis/lambda-2015-03-31.normal.json
lambda_svc <- lambda()
b3 <- tryCatch(lambda_svc$get_alias(FunctionName = "aslkfjalksfjalskfjalskfjlaksjflaksjflkasjflkjs", Name = "alsfkjaslkfj"), error = function(e) e)
b3$message
class(b3)
attributes(b3)

# json - dynamodb
# https://github.com/aws/aws-sdk-js/blob/master/apis/dynamodb-2012-08-10.normal.json
dynamo_svc <- dynamodb()
b4 <- tryCatch(dynamo_svc$describe_table(TableName = "asfahsfkljaslfkjaslkfjalkfjasl" ), error = function(e) e)
b4$message
class(b4)
attributes(b4)

# query - sts
# https://github.com/aws/aws-sdk-js/blob/master/apis/sts-2011-06-15.normal.json
sts_svc <- sts()
b5 <- tryCatch(sts_svc$get_access_key_info(AccessKeyId = "alskfjaskfjalskfjlaskjf"), error = function(e) e)
b5$message
class(b5)
attributes(b5)

# rest-xml - cloudfront
# https://github.com/aws/aws-sdk-js/blob/master/apis/cloudfront-2019-03-26.normal.json
cloudfront_svc <- cloudfront()
b6 <- tryCatch(cloudfront_svc$get_distribution(Id = "alskfjaskfjalskfjlaskjf"), error = function(e) e)
b6$message
class(b6)
attributes(b6)
```

### CODE REVIEW

**Comment by davidkretch** [comment]
Awesome!  We'll look in the next couple days.  Thank you!

**Comment by ryanb8** [comment]
This addresses #245 !

File: paws.common/tests/testthat/test_handlers_ec2query.R [hunk 1]
```diff
@@ -411,3 +412,18 @@ test_that("unmarshal timestamp", {
   out <- req$data
   expect_equal(out$Timestamp, unix_time(0))
 })
+
+test_that("unmarshall error", {
```
Comment (davidkretch) [inline paws.common/tests/testthat/test_handlers_ec2query.R]: There is a typo here: "unmarshall" -> "unmarshal"

**Comment by davidkretch** [comment]
This is really nice.  I like this a lot.  My comments:

1. I think variables should be camel_case, unless it is a name provided in the error response itself (e.g. BucketName).  In general I think we use snake_case for all package-internal names, and then keep AWS's names for all inputs and outputs, which are almost always UpperCamelCase.

2. I think `http_error` should be something like `aws_error` since it's more specific than an HTTP error.

3. A request can fail before it gets sent to AWS in which case it will stop with an error but the status code will be NA.  I think in these cases, `http_error` probably should avoid setting class `http_NA`.

2. I think it makes sense to include the S3 error example in the examples folder, but this is really up to you.

File: paws.common/R/error.R [hunk 2]
```diff
@@ -2,14 +2,56 @@
...
 )
 
-# Creates a string for printing for a given Error object.
-format_error <- function(e) {
-  s <- e$code
-  if (!is.null(e$message) && e$message != "") {
-    s <- sprintf("%s: %s", e$code, e$message)
+#' Generate a classed http error
+#'
+#' This function generate S3 error objects which are passed to
```
Comment (davidkretch) [inline paws.common/R/error.R]: Typo: "generate" -> "generates"

File: paws.common/R/error.R [hunk 3]
```diff
@@ -2,14 +2,56 @@
...
+#' [stop()] to generate classes errors.
```
Comment (davidkretch) [inline paws.common/R/error.R]: "classes" -> "classed"

File: paws.common/R/error.R [hunk 4]
```diff
@@ -2,14 +2,56 @@
...
+#' These can be used in conjunction with [tryCatch()]
+#' to respond differently to different type of failure.
```
Comment (davidkretch) [inline paws.common/R/error.R]: "type" -> "types"

**Comment by ryanb8** [comment]
I agree with all of your comments. I will work on making those changes in the next 7ish days.

**Comment by ryanb8** [comment]
Ok, I *think* this is good to go now. I need to rebuild paws so I can fully test the added example. I will test this first thing tomorrow morning.

All the paws.common checks pass!

Changes above:
1. Fixed merge conflict (twice... 😖 )
2. Fixed all typos as addressed
3. Renamed `http_error` to `aws_error`
4. Made `aws_error` not set error class of `http_NA_integer_`
5. Added unit test for `aws_error` to address 4) and general case
6. Removed roxygen export tag of `aws_error` (this seems to follow precedent within the paws.common package)
7. Switched to snake_case where appropriate.

If desired, I can squash the above 4-5 commits and force push for final review, or we can squash-merge, or we can not worry about it. Your call!

**Comment by davidkretch** [comment]
Awesome!  I'll look tomorrow.  Don't worry about squashing.  👍

**Comment by ryanb8** [comment]
It looks like all my manual examples (very top) and example/error_handling.R work correctly. This is good-to-go from my perspective.

File: paws.common/tests/testthat/test_handlers_query.R [hunk 5]
```diff
@@ -708,34 +708,44 @@ test_that("unmarshal enums", {
 request <- list()
 
 test_that("unmarshal error", {
-  data <- "<ErrorResponse><Error><Code>FooError</Code><Message>Foo</Message><RequestId>123</RequestId><HostId>ABC</HostId></Error></ErrorResponse>"
+  data <- "<error_response><Error><Code>FooError</Code><Message>Foo</Message><RequestId>123</RequestId><HostId>ABC</HostId></Error></error_response>"
```
Comment (davidkretch) [inline paws.common/tests/testthat/test_handlers_query.R]: The only thing I disagree with I think this should stay ErrorResponse.  We don't control the response, we just receive it, so I think the example response should use the actual response layout.  The name doesn't affect the result.

File: paws.common/tests/testthat/test_handlers_query.R [hunk 6]
```diff
@@ -708,34 +708,44 @@ test_that("unmarshal enums", {
...
   out <- request$error
   expect_equal(out$code, "FooError")
   expect_equal(out$message, "Foo")
+  expect_equal(out$status_code, 400)
+  expect_equal(out$error_response$RequestId, "123")
 })
 
 test_that("unmarshal error with an empty message", {
-  data <- "<ErrorResponse><Error><Code>FooError</Code><Message></Message><RequestId>123</RequestId><HostId>ABC</HostId></Error></ErrorResponse>"
+  data <- "<error_response><Error><Code>FooError</Code><Message></Message><RequestId>123</RequestId><HostId>ABC</HostId></Error></error_response>"
```
Comment (davidkretch) [inline paws.common/tests/testthat/test_handlers_query.R]: This is the one other example.

**Comment by davidkretch** [comment]
Hey, thank you!  Everything looks really good.  I just had one thing to change in the `query_unmarshal_error` unit tests, which I can take care of if you're ok with that.

The other thing that I want to take another day to understand is why it seems like I have to re-install the paws packages to get some of the manual examples to work correctly, at least on my machine.  I don't think this is an issue with the error functions, rather with the relationship between paws.common and the other paws packages, but I want to look deeper.

**Comment by ryanb8** [comment]
Yep, feel free to change anything you'd like. No need to chat with me about it or what not.

Regarding the manual examples, yes I had to rebuild everything to get my manual tests to work. I thought that seemed strange at first, but I haven't really looked at how the api is read and the code is generated during the build, so I was unsure if that was expected behavior or not. I had the thought that it was a my machine thing or a result of me having lots of R sessions going at once with loaded libraries.

File: paws.common/tests/testthat/test_handlers_query.R (hunk 5 above)
Comment (ryanb8) [reply paws.common/tests/testthat/test_handlers_query.R]: Yep. I completely agree. I'll make these changes.

**Comment by ryanb8** [comment]
I think I fixed the things you were going to fix. I didn't put two-and-two together until it was too late. Still, feel free to change whatever!

**Comment by davidkretch** [comment]
Looks good to me!

It will be a few weeks before we get the next version of paws.common and paws on CRAN though.  They have limits on how frequently we submit.

With respect to the build issue, we're going to fix that.  The goal is that changes like this only require updating paws.common.  I'm glad we came across it now.

**Comment by ryanb8** [comment]
Huzzah!

**Pull request merged**